src/RobustLeastSquares.jl

```diff
@@ -1,5 +1,307 @@
+    w[est.rng] = m_estimator_sqrtweight(r[est.rng],est.est)
+    return w
+end
+
+
+"""
+A custom type for applying two estimators to different ranges
+(uses L2 for any unspecified range)
+"""
+immutable DualEstimatorConvex{T1<:MEstimatorConvex,T2<:MEstimatorConvex} <: MEstimatorConvex
+    rng1::UnitRange{Int64}
+    est1::T1
+    rng2::UnitRange{Int64}
+    est2::T2
+end
+immutable DualEstimator{T1<:MEstimator,T2<:MEstimator} <: MEstimatorNonConvex
+    rng1::UnitRange{Int64}
+    est1::T1
+    rng2::UnitRange{Int64}
+    est2::T2
+end
+#DualEstimator = union(DualEstimatorConvex,DualEstimatorNonConvex)
+function m_estimator_rho{T1,T2}(r,est::Union{DualEstimatorConvex{T1,T2},DualEstimator{T1,T2}})
+    rho = 0.5*r.^2
+    rho[est.rng1] = m_estimator_rho(r[est.rng1],est.est1)
+    rho[est.rng2] = m_estimator_rho(r[est.rng2],est.est2)
+    return rho
+end
+function m_estimator_psi{T1,T2}(r,est::Union{DualEstimatorConvex{T1,T2},DualEstimator{T1,T2}})
+    psi = r
+    psi[est.rng1] = m_estimator_psi(r[est.rng1],est.est1)
+    psi[est.rng2] = m_estimator_psi(r[est.rng2],est.est2)
+    return phi
+end
+function m_estimator_weight{T1,T2}(r,est::Union{DualEstimatorConvex{T1,T2},DualEstimator{T1,T2}})
+    w = ones(size(r))
+    w[est.rng1] = m_estimator_weight(r[est.rng1],est.est1)
+    w[est.rng2] = m_estimator_weight(r[est.rng2],est.est2)
+    return w
+end
+function m_estimator_sqrtweight{T1,T2}(r,est::Union{DualEstimatorConvex{T1,T2},DualEstimator{T1,T2}})
+    w = ones(size(r))
+    w[est.rng1] = m_estimator_sqrtweight(r[est.rng1],est.est1)
+    w[est.rng2] = m_estimator_sqrtweight(r[est.rng2],est.est2)
+    return w
+end
+
+"Recreate an M-Estimator of the same type using the MAD (median absolute deviation) of the residual"
+mestimator_mad{T<:MEstimator}(Est::Type{T}, res, factor=1.0) = Est(factor*1.43*StatsBase.mad(res))
+function mestimator_mad{T}(Est::Union{Type{SubsetEstimator{T}},Type{SubsetEstimatorConvex{T}}}, rng, res, factor=1.0)
+    return Est(rng, T(factor*1.43*StatsBase.mad(res[rng])))
+end
+function mestimator_mad{T1,T2}(Est::Union{Type{DualEstimator{T1,T2}},Type{DualEstimatorConvex{T1,T2}}}, rng1, rng2, res, factor=1.0)
+    return Est(rng1, T1(factor*1.43*StatsBase.mad(res[rng1])), rng2, T2(factor*1.43*StatsBase.mad(res[rng2])))
+end
+
+mestimator_mad(est::MEstimator, res, factor=1.0) = typeof(est)(factor*1.43*StatsBase.mad(res))
+function mestimator_mad{T}(est::Union{SubsetEstimator{T},SubsetEstimatorConvex{T}}, res, factor=1.0)
+    return typeof(est)(est.rng, T(factor*1.43*StatsBase.mad(res[est.rng])))
+end
+function mestimator_mad{T1,T2}(est::Union{DualEstimator{T1,T2},DualEstimatorConvex{T1,T2}}, res, factor=1.0)
+    return typeof(est)(est.rng1, T1(factor*1.43*StatsBase.mad(res[est.rng1])), est.rng2, T2(factor*1.43*StatsBase.mad(res[est.rng2])))
+end
+
+
+
+
+
+function solve(A,b,weights=ones(length(b)),method=:qr,x0=nothing)
+    if method == :qr
+        return (spdiagm(weights)*A) \ (spdiagm(weights)*b)
```

Comment:
Also we should ideally support sparse and dense matrices equally well.  The simple way to do this would be using `Diagonal` rather than `spdiagm`, but unfortunately `*(::Diagonal{Int64}, ::SparseMatrixCSC{Float64,Int64})` apparently isn't defined in base (maybe we should just define it above as a temp hack?  Not sure how easy this is but it should be pretty straightforward.)

Unfortunately